File: backend/src/main/kotlin/com/migrationtimeline/analysis/analyse.kt
package com.migrationtimeline.analysis

import com.migrationtimeline.models.*

internal data class AnalysisContext(
    val partitionedTables: Set<String>
)

fun analyse(migrations: List<Migration>): AnalysisResult {
    val partitionedTables = migrations
        .flatMap { it.operations }
        .filterIsInstance<CreateTable>()
        .filter { it.isPartitioned }
        .map { it.tableName }
        .toSet()

    val context = AnalysisContext(partitionedTables)

    val warnings = migrations.flatMap { migration ->
        migration.operations.flatMap { analyzeOperation(it, context) }
    }

    return AnalysisResult(warnings)
}

private fun analyzeOperation(op: Operation, context: AnalysisContext): List<Warning> = listOfNotNull(
    detectAccessExclusiveLock(op, context)
)

File: backend/internal/analysis/analyse.go
package analysis

import "migration-timeline/backend/internal/models"

type analysisContext struct {
	partitionedTables map[string]bool
}

func Analyse(migrations []*models.Migration) *models.AnalysisResult {
	partitionedTables := make(map[string]bool)

	for _, m := range migrations {
		for _, stmt := range m.Statements {
			for _, op := range stmt.Operations {
				if ct, ok := op.(models.CreateTable); ok && ct.IsPartitioned {
					partitionedTables[ct.TableName] = true
				}
			}
		}
	}

	ctx := &analysisContext{partitionedTables: partitionedTables}
	var warnings []models.Warning

	for _, m := range migrations {
		for _, stmt := range m.Statements {
			if w := analyzeStatement(stmt, m.ID, ctx); w != nil {
				warnings = append(warnings, w)
			}
		}
	}

	return &models.AnalysisResult{
		Migrations: migrations,
		Warnings:   warnings,
	}
}

func analyzeStatement(stmt models.Statement, migrationID string, ctx *analysisContext) models.Warning {
	return detectAccessExclusiveLock(stmt, migrationID, ctx)
}
